'Classify' (Choropleth) tool for layer styles

mbertrand opened the pull request:
Choropleth style generator for map layers.  This depends on GeoServer's sldService community module (sldService-2.4-SNAPSHOT.jar, xom-1.1.jar need to be added to GeoServer's WEB-INF/lib directory).  

When the WMSStylesDialog panel is brought up for a layer, an ajax call is made to see if the sldService module is installed, and displays a 'Classify' toolbar/button only if it receives a valid response.  The ajax call requires a new url for GeoNode's proxy app ("gs/rest/sldservice").

Pressing the Classify button brings up a modified RulePanel, with classification options at the bottom (Attribute, Method, Classes, Color Ramp, Reverse Colors).  The styles are generated when the Apply button at bottom is pressed.

Screenshots: http://imgur.com/a/IbswZ

This tool also depends on a patch to GeoTools/GeoServer which made it into trunk but may not be in the 2.4.x and earlier branches: http://jira.codehaus.org/browse/GEOS-6150, https://github.com/geotools/geotools/pull/324. Without this fix, the classifier will generate error tiles if it creates style rules for integer fields containing decimals (for example, a value range of 2.0 - 10.1)

bartvde commented:
@mbertrand why have the spinner classes been pulled into the gxp namespace? Can we not treat them like we treat other ux modules?

mbertrand commented:
@bartvde I will look into that.  I haven't touched this functionality in at least a year and probably should check if all the dependent parts still work together (such as the GeoServer sldService community module, which I last used with version 2.3).  I'll try to get to it this week sometime.

bartvde commented:
cool I got pinged on some geonode e-mail / ticket I had also forgotten about this
happy to review this once you've confirmed it still works.

capooti commented:
looking forward to have this working in GN guys :) thanks a lot in advance

mbertrand commented:
Hi @bartvde, I changed the namespace for the Spinner classes back to Ext.ux (not sure why I changed it to gxp earlier).  This is where I got them from originally: http://dev.sencha.com/deploy/ext-3.4.0/examples/spinner/spinner.html

@capooti, I tested it out with geonode and it works.

bartvde commented:
@mbertrand we don't manage Ext ux components at the GXP level, but at the application level. So GeoExplorer will need to get the Ext ux components in here: https://github.com/boundlessgeo/suite/tree/master/geoexplorer/app/static/script/ux in a subsequent PR. Also if you have a GXP example, it will need to be added there (https://github.com/boundlessgeo/gxp/tree/master/examples/ux/treegrid).

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+
+    rulePanel: null,
+
+    classifyText: "Classify",
```
please also add least add this to the locale files (https://github.com/boundlessgeo/gxp/tree/master/src/script/locale), for the other languages just copy in the English language.

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+ */
+
+Ext.namespace("gxp");
+
```
Please add at least some API docs, it doesn't have to be much, but something at least.

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                    id: "choropleth_color_start",
+                    name: "color_start",
+                    fieldLabel: this.colorStartText,
+                    emptyText: OpenLayers.Renderer.defaultSymbolizer.strokeColor,
```
should be @require'd then

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                    listeners: {
+                        valid: function(field) {
+                            this.rulePanel.rule[field.name] = field.getValue();
+                            //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
can this be removed?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                    listeners: {
+                        valid: function(field) {
+                            this.rulePanel.rule[field.name] = field.getValue();
+                            //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
same comment, instead of commenting out please remove

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        },
+                        scope: this
+                    },
+                    scope: this
```
why do you need scope at  two levels?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+
+
+        var classNumSelector = new Ext.ux.form.SpinnerField({
+            fieldLabel: 'Classes',
```
should be i18n in the prototype?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            listeners: {
+                'change':function(spinner, value){
+                    this.rulePanel.rule['intervals'] = value;
+                    //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
please remove instead of commenting

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            width: 110,
+            listeners: {
+                'change':function(spinner, value){
+                    this.rulePanel.rule['intervals'] = value;
```
can you get the name from the spinner instead of hard-coding 'intervals'  again?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+        var colorDropdown = new Ext.form.ComboBox({
+            id: 'choropleth_colorramp',
+            name: 'ramp',
+            fieldLabel: 'Color Ramp',
```
i18n needed

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            name: 'ramp',
+            fieldLabel: 'Color Ramp',
+            store:  new Ext.data.ArrayStore({
+                id: 0,
```
is this needed?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        case "Blue":
+                            this.rulePanel.rule["color_start"] = "#f7fbff";
+                            this.rulePanel.rule["color_end"] = "#08306b";
+                            //this.rulePanel.rule["color_mid"] = "#fd8d3c";
```
please remove commented code

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        case "Red":
+                            this.rulePanel.rule["color_start"] = "#fff5f0";
+                            this.rulePanel.rule["color_end"] = "#67000d";
+                            //this.rulePanel.rule["color_mid"] = "";
```
please remove commented code

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+
+                    switch(cmb.value) {
+                        case "Blue":
+                            this.rulePanel.rule["color_start"] = "#f7fbff";
```
should these colors be configurable or not needed?

mbertrand replied:
Custom colors can be chosen by selecting the 'Custom' option in dropdown.

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        case "Orange":
+                            this.rulePanel.rule["color_start"] = "#fff5eb";
+                            this.rulePanel.rule["color_end"] = "#f16913";
+                            //this.rulePanel.rule["color_mid"] = "";
```
please remove commented code

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            listeners: {
+                'select': function(cmb, data, idx) {
+                    //If Custom: display start, end, middle color picker;
+                    colorPanel.setVisible(cmb.value == "Custom");
```
Can we make "Custom" a property of this class, so it doesn't have to get repeated all the time?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                            //this.rulePanel.rule["color_start"] = "#5e4fa2";
+                            //this.rulePanel.rule["color_end"] = "#9e0142";
+                            //this.rulePanel.rule["color_mid"] = "#ffffbf";
+                            this.rulePanel.rule[cmb.name] = "Jet";
```
please remove commented code

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                            this.rulePanel.rule[cmb.name] = "Jet";
+                            break;
+                        default:
+                            this.rulePanel.rule["color_mid"] = "";
```
is color_mid even used anywhere?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+        var methodDropdown = new Ext.form.ComboBox({
+            id: 'choropleth_method',
+            name: 'method',
+            fieldLabel: 'Method',
```
needs i18n

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            name: 'method',
+            fieldLabel: 'Method',
+            store:  new Ext.data.ArrayStore({
+                id: 0,
```
is this even a config in ExtJS?

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+            mode: 'local',
+            width: 110,
+            editable: false,
+            emptyText: 'Select method',
```
needs i18n

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                    //If uniqueInterval: disable # classes
+                    classNumSelector.setDisabled(cmb.value == "uniqueInterval");
+                    this.rulePanel.rule[cmb.name] = cmb.value;
+                    //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
please remove commented code

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+
+                    methodDropdown.clearValue();
+                    var methodStore =   methodDropdown.getStore();
+                    for (var i = 0; i < methodStore.data.length; i++) {
```
Please use:

```
for (var i = 0, ii=methodStore.data.length; i <ii ; i++) 
```

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        ['jenks', this.methodJenksText]]
+                    );
+
+                    //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
please remove

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+                        handler: function(item, e)
+                        {
+                            this.rulePanel.rule['reverse'] = item.checked;
+                            //this.rulePanel.fireEvent("change", this.rulePanel, this.rulePanel.rule);
```
please remove

bartvde commented on src/script/widgets/ClassificationPanel.js:
```diff
@@ -0,0 +1,317 @@
+
+});
+
+/** api: xtype = gxp_linesymbolizer */
```
needs an update

mbertrand replied:
Not sure I understand, what needs to be updated?  The xtype?

bartvde commented on src/script/widgets/RulePanel.js:
```diff
@@ -489,6 +500,29 @@ gxp.RulePanel = Ext.extend(Ext.TabPanel, {
+            this.rule["classify"] = true;
+        }
+
+        return new gxp.ClassificationPanel({
```
this needs a @require at the top of this file?

mbertrand replied:
Added  @include for widgets/ClassificationPanel.js

bartvde commented on src/script/widgets/RulePanel.js:
```diff
@@ -413,6 +423,7 @@ gxp.RulePanel = Ext.extend(Ext.TabPanel, {
                     width: 150,
                     items: [{
                         xtype: "textfield",
+                        hidden: this.classifyEnabled,
```
this property should be documented in this class?

mbertrand replied:
Done

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -20,6 +20,21 @@
+ * @param obj
+ * @constructor
+ */
+OpenLayers.Format.Filter.v1_1_0.prototype.readers.ogc.Literal = function(node, obj) {
```
Needs a @require

mbertrand replied:
Added @require for OpenLayers/Format/Filter/v1_1_0.js

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -49,6 +64,10 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
     /** api: config[chooseStyleText] (i18n) */    
     chooseStyleText: "Choose style",
+    /** api: config[classifyStyleText] (i18n) */
+    classifyStyleText:"Classify",
```
needs to be added to the locale files as well in the locale dir

mbertrand replied:
Done (noticed that WMSStylesDialog section was missing for French file; added it in)

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -382,7 +406,8 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
                     styleProperties.destroy();
-                }
+                },
+                scope: this
```
scope is not needed here right?

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -415,6 +441,83 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
     
+
+    /** api: method[classifyStyleRules]
+     *  :arg prevStyle: ``Ext.data.Record``
```
there is no argument?

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -415,6 +441,83 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
+
+        var rule = userStyle.rules[0];
+        var origRules = [];
+        for (var i=0; i < userStyle.rules.length; i++){
```
again, optimize the for loop

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -415,6 +441,83 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
+                rule: rule,
+                fonts:this.fonts,
+                classifyEnabled: true,
+                attributes: new GeoExt.data.AttributeStore({
```
since this is now duplicated code, can you make a function that creates the attribute store instead an reuse that in the two places?

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -647,13 +751,75 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
+                    var xmlParser = new OpenLayers.Format.XML();
+
+                    var xmlRules = xmlParser.getElementsByTagNameNS(xmlParser.read(result.responseText).documentElement, "*", "Rule");
+                    for (var i = 0; i < xmlRules.length; i++) {
```
please optimise for loops

bartvde commented on src/script/widgets/WMSStylesDialog.js:
```diff
@@ -925,7 +1098,62 @@ gxp.WMSStylesDialog = Ext.extend(Ext.Container, {
+            url = layerRecord.get("restUrl");
+        }
+        if (!url) {
+            url = layer.url.split("?").shift().replace(/\/(wms|ows)\/?$/, "/rest");
```
can this code path be re-used, it's now duplicated

bartvde commented on src/theme/all.css:
```diff
@@ -605,3 +605,58 @@ div.olMap div.gx-overlay-playback {
 }
+.x-form-spinner-proxy{
+    /*background-color:#ff00cc;*/
+}
```
can be removed?

bartvde commented on src/theme/all.css:
```diff
@@ -605,3 +605,58 @@ div.olMap div.gx-overlay-playback {
+    background:transparent url('img/spinner.gif') no-repeat 0 0;
+    right: 0px;
+}
+.x-form-field-wrap .x-form-spinner-overup{
```
I'd leave it up to the applications to include the spinner css in their builds instead of duplicating it here.

bartvde commented:
@mbertrand I've finished my review, TIA for addressing the comments.

mbertrand commented:
Thanks @bartvde I'll start working on making these changes.

capooti commented:
@bartvde is this ready to be merged, or something is still missing? Thanks

bartvde commented:
@capooti merged, I think I had simply forgotten about this one

capooti commented:
awesome :)
thanks @bartvde
